New resources: `azurerm_lighthouse_definition` and 'azurerm_lighthouse_assignment'

Humoiz opened the pull request:
Implementing features mentioned in the feature request: #3941

This PR adds following new resources and its documentation:
azurerm_lighthouse_definition
azurerm_lighthouse_assignment

Data source:
azurerm_lighthouse_definition

The resources in this PR needs multiple tenant to run the scenario. Tenant ID and Object ID of the user, usergroup or service principal is needed from the second tenant. I ran the tests after setting the environment variable ARM_TENANT_ID_ALT & ARM_PRINCIPAL_ID_ALT_TENANT locally. 

Here is the azure docs which explains how to setup these resources:
https://docs.microsoft.com/en-us/azure/lighthouse/how-to/onboard-customer#gather-tenant-and-subscription-details

All the tests are passing:
```
$  TF_ACC=1 go test -v ./azurerm/internal/services/lighthouse/tests/ -run='(TestAccDataSourceAzureRMLighthouseDefinition)'
=== RUN   TestAccDataSourceAzureRMLighthouseDefinition_basic
=== PAUSE TestAccDataSourceAzureRMLighthouseDefinition_basic
=== CONT  TestAccDataSourceAzureRMLighthouseDefinition_basic
--- PASS: TestAccDataSourceAzureRMLighthouseDefinition_basic (39.53s)
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/lighthouse/tests    41.870s
```
```
$  TF_ACC=1 go test -v ./azurerm/internal/services/lighthouse/tests/ -run='(TestAccAzureRMLighthouseAssignment)'
=== RUN   TestAccAzureRMLighthouseAssignment_basic
=== PAUSE TestAccAzureRMLighthouseAssignment_basic
=== RUN   TestAccAzureRMLighthouseAssignment_requiresImport
=== PAUSE TestAccAzureRMLighthouseAssignment_requiresImport
=== RUN   TestAccAzureRMLighthouseAssignment_emptyID
=== PAUSE TestAccAzureRMLighthouseAssignment_emptyID
=== CONT  TestAccAzureRMLighthouseAssignment_basic
=== CONT  TestAccAzureRMLighthouseAssignment_emptyID
=== CONT  TestAccAzureRMLighthouseAssignment_requiresImport
--- PASS: TestAccAzureRMLighthouseAssignment_emptyID (57.84s)
--- PASS: TestAccAzureRMLighthouseAssignment_basic (58.12s)
--- PASS: TestAccAzureRMLighthouseAssignment_requiresImport (70.02s)
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/lighthouse/tests    72.475s

```
```
$  TF_ACC=1 go test -v ./azurerm/internal/services/lighthouse/tests/ -run='(TestAccAzureRMLighthouseDefinition)'
=== RUN   TestAccAzureRMLighthouseDefinition_basic
=== PAUSE TestAccAzureRMLighthouseDefinition_basic
=== RUN   TestAccAzureRMLighthouseDefinition_requiresImport
=== PAUSE TestAccAzureRMLighthouseDefinition_requiresImport
=== RUN   TestAccAzureRMLighthouseDefinition_complete
=== PAUSE TestAccAzureRMLighthouseDefinition_complete
=== RUN   TestAccAzureRMLighthouseDefinition_update
=== PAUSE TestAccAzureRMLighthouseDefinition_update
=== RUN   TestAccAzureRMLighthouseDefinition_emptyID
=== PAUSE TestAccAzureRMLighthouseDefinition_emptyID
=== CONT  TestAccAzureRMLighthouseDefinition_basic
=== CONT  TestAccAzureRMLighthouseDefinition_update
=== CONT  TestAccAzureRMLighthouseDefinition_requiresImport
=== CONT  TestAccAzureRMLighthouseDefinition_complete
=== CONT  TestAccAzureRMLighthouseDefinition_emptyID
--- PASS: TestAccAzureRMLighthouseDefinition_emptyID (35.96s)
--- PASS: TestAccAzureRMLighthouseDefinition_complete (37.20s)
--- PASS: TestAccAzureRMLighthouseDefinition_update (59.27s)
--- PASS: TestAccAzureRMLighthouseDefinition_basic (68.36s)
--- PASS: TestAccAzureRMLighthouseDefinition_requiresImport (87.92s)
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/lighthouse/tests    90.403s
```

Humoiz commented on azurerm/internal/services/managedservices/tests/data_source_registration_definition_test.go:
```diff
@@ -0,0 +1,72 @@
+	data := acceptance.BuildTestData(t, "data.azurerm_registration_definition", "test")
+	// Multiple tenants are needed to test this resource. 
+	// Second tenant ID needs to be set as a enviornment variable ARM_SECOND_TENANT_ID.
+	secondTenantID := os.Getenv("ARM_SECOND_TENANT_ID")
```
The resources in this PR needs multiple tenant to run the scenario. I ran these tests after setting the environment variable ARM_SECOND_TENANT_ID. Is there a way to setup the second tenant for these tests?

Humoiz commented on vendor/github.com/Azure/azure-sdk-for-go/services/managedservices/mgmt/2019-06-01/managedservices/client.go:
```diff
@@ -0,0 +1,50 @@
+// Package managedservices implements the Azure ARM Managedservices service API version 2019-06-01.
+//
+// Specification for ManagedServices.
```
I copied the SDK files manually. Is there a way to automatically copy these files?

katbyte replied:
Once you add the imports `go mod vendor` should do it

katbyte commented:
@Humoiz,

It's not entire clear what a `registration definition` applies to, i think as the ticket suggested `azurerm_lighthous_*` would be better. WDYT?

Humoiz commented on website/docs/r/registration_assignment.html.markdown:
```diff
@@ -0,0 +1,70 @@
+
+---
+
+# azurerm_registration_definition
```
I will change this to azurerm_registration_assignment in the next round of changes.

Humoiz commented:
@katbyte - This PR is ready for your review. Please provide your feedback when time permits.

Humoiz commented:
@katbyte - Just checking if I can answer any questions you may have.

katbyte commented:
@Humoiz - the SDK & API often leave much to be desired in their clarity of naming. Could we update these to all use lighthouse including the service package name? thanks!

Humoiz commented:
> @Humoiz - the SDK & API often leave much to be desired in their clarity of naming. Could we update these to all use lighthouse including the service package name? thanks!

I have renamed from registration_* to lighthouse_*, also updated the function names, etc. Let me know how does it looks to you now.

Humoiz commented:
@katbyte -I have addressed your initial comments. I need to use these new resources in about two weeks. Can you please provide your feedback?

Humoiz commented:
@katbyte -The build is failing and I am not sure how to fix it. I tried 'go mod vendor' command and it deletes several folders, not sure if I am missing anything. Any pointers?

katbyte commented:
@Humoiz - i merged master and fixed up the vendor folder 🙂

Humoiz commented:
> @Humoiz - i merged master and fixed up the vendor folder 🙂

Thank you!! :)

ArcturusZhang commented on azurerm/internal/services/managedservices/resource_arm_lighthouse_assignment.go:
```diff
@@ -0,0 +1,194 @@
+		},
+
+		Schema: map[string]*schema.Schema{
+			"registration_assignment_id": {
```
I think we could just name this as `name` like `azurerm_policy_assignment` does. A resource itself would have an ID, and naming this as some ID may cause confusion. What do you think?

Humoiz replied:
@ArcturusZhang  I reviewed the api again and I agree with naming it from `registration_assignment_id` to `name`. Is there any other feedback which I can address along with this change?

Humoiz replied:
I have changed this to `name`. It can't be anything other than uuid.

Humoiz commented on azurerm/internal/services/managedservices/data_source_lighthouse_definition.go:
```diff
@@ -0,0 +1,107 @@
+		},
+
+		Schema: map[string]*schema.Schema{
+			"registration_definition_id": {
```
@ArcturusZhang @katbyte - Do you think that registration_definition_id and registration_definition_name should be renamed to lighthouse_definition_id and lighthouse_definition_name to match the resource name azurerm_lighthouse_dinition?

katbyte replied:
thats a good point! yes anytime we are referrring to another resources name property or id we should be explicit on what resource type we are referring to

Humoiz replied:
Sure. I will change this as well to lighthouse_definition_id and lighthouse_definition_name

katbyte commented on website/allowed-subcategories:
```diff
@@ -36,6 +36,7 @@ Logic App
 Machine Learning
 Maintenance
 Managed Applications
+Managed Services
```
Lighthouse

katbyte reviewed (dismissed):
Thanks @Humoiz, i;ve left some comments mostly around naming - otherwise this is looking good and should get into the release next week!

Humoiz commented on azurerm/internal/services/managedservices/tests/data_source_lighthouse_definition_test.go:
```diff
@@ -0,0 +1,66 @@
+
+  authorization {
+    principal_id       = "%s"
+    role_definition_id = "b24988ac-6180-42a0-ab88-20f7382dd24c"
```
This is azure built-in role "Contributor". This service supports azure build in roles. I will also place a comment in the test saying that it is a build in Contributor role.

https://docs.microsoft.com/en-us/azure/role-based-access-control/built-in-roles#all

Humoiz replied:
I have updated the code to use the data source to get the role and than pass the reference in the role_definition_id for better readability.

`data "azurerm_role_definition" "contributor" {
  role_definition_id = "b24988ac-6180-42a0-ab88-20f7382dd24c"
}`

Humoiz commented on website/docs/r/lighthouse_definition.html.markdown:
```diff
@@ -0,0 +1,70 @@
+
+The following arguments are supported:
+
+* `registration_definition_id` - (Optional) A unique UUID/GUID which identifies this lighthouse definition - one will be generated if not specified. Changing this forces a new resource to be created.
```
There are two properties in `azurerm_lighthouse_definition` resource.

`registration_definition_id` & `registration_definition_name`.

If I change registration_definition_id to `name`. Dont think it be confusing to have two properties such as `name` & `lighthouse_definition_name`.  WDYT?

The api documentation says that it should be a GUID. I haven`t tested a non-guid but will do.

Humoiz replied:
I think `uuid` can be confusing. 

This is what I suggest. 

Changing `registration_definition_name` to `name` and changing `registration_definition_id` to `lighthouse_definition_id`. In this case we can do our best to document `lighthouse_definition_id` and `id` clearly. Let me know what do you think.

katbyte replied:
THis works for me!

Humoiz replied:
ok great. I have pushed all the changes.

Humoiz commented on website/docs/r/lighthouse_assignment.html.markdown:
```diff
@@ -0,0 +1,69 @@
+
+The following arguments are supported:
+
+* `registration_assignment_id` - (Optional) A unique UUID/GUID which identifies this lighthouse assignment- one will be generated if not specified. Changing this forces a new resource to be created.
```
I`m changing it to `name.`

Humoiz commented:
> Thanks @Humoiz, i;ve left some comments mostly around naming - otherwise this is looking good and should get into the release next week!

Thank you very much for your feedback, much appreciated. I have resolved all your feedback and currently testing the new changes. I`ll push my changes tomorrow.

Humoiz commented on azurerm/internal/services/managedservices/tests/resource_arm_lighthouse_assignment_test.go:
```diff
@@ -0,0 +1,213 @@
+}
+
+resource "azurerm_lighthouse_definition" "test" {
+  registration_definition_name = "acctestrd-%d"
```
Updated all references.

logachev commented:
@katbyte do you have any updates on this one?

katbyte commented:
@logachev - i believe its waiting on #6809 for binary testing so we can use multiple provider blocks in acctests to correctly test across tenants

Humoiz commented:
> @logachev - i believe its waiting on #6809 for binary testing so we can use multiple provider blocks in acctests to correctly test across tenants

@katbyte - Do we have an ETA for this to be merged? #6809

tombuildsstuff commented:
👋 

Just to give an update here, whilst this PR was dependent on #6809 - we ended up disabling Binary Testing in that PR - work to enable Binary testing will happen in a subsequent PR (and is being investigated) - however unfortunately this PR is still blocked for the moment.

Thanks!

lukiffer commented:
@tombuildsstuff is there an open forum where the binary testing piece is being discussed/investigated? This is an important feature for us, and we'd be happy to contribute.

Also, could you provide a bit more background on which parts of the platform SDK are requiring the addition of `TestMain` in each package, and moreover what the issue was with the previous implementation of conditionally enabling it for packages that needed it?

WodansSon commented:
@Humoiz, the Travis CI is currently failing with:

```
==> Checking source code with go mod vendor...
 .../managedservices/mgmt/2019-06-01/managedservices/operations.go     | 1 -
 .../mgmt/2019-06-01/managedservices/registrationassignments.go        | 4 ----
 .../mgmt/2019-06-01/managedservices/registrationdefinitions.go        | 4 ----
 3 files changed, 9 deletions(-)

Unexpected difference in vendor/ directory. Run 'go mod vendor' command or revert any go.mod/go.sum/vendor changes and commit. 
```

Can you please verify if the vendor changes are needed? Do you need to pull the current `master` branch into your fork?

Humoiz commented:
> @Humoiz, the Travis CI is currently failing with:
> 
> ```
> ==> Checking source code with go mod vendor...
>  .../managedservices/mgmt/2019-06-01/managedservices/operations.go     | 1 -
>  .../mgmt/2019-06-01/managedservices/registrationassignments.go        | 4 ----
>  .../mgmt/2019-06-01/managedservices/registrationdefinitions.go        | 4 ----
>  3 files changed, 9 deletions(-)
> 
> Unexpected difference in vendor/ directory. Run 'go mod vendor' command or revert any go.mod/go.sum/vendor changes and commit. 
> ```
> 
> Can you please verify if the vendor changes are needed? Do you need to pull the current `master` branch into your fork?

I just merged master to resolve the conflicts. I dont understand how go mod vendor works and when I run this command there are several files sgows up with changes. I am not sure what needs to be done here to fix the build failure.

jackofallops commented:
Hi @Humoiz 
I hope you don't mind, but I've resolved a merge conflict and rebased to bring the branch up to date to help us assess if this can go in manually tested ahead of the changes we're working on for the binary testing capability that's imminently due.  I'll be focused on this work today and I'm hoping we'll be able to get it in for the next release.

tombuildsstuff commented on azurerm/internal/services/lighthouse/resource_arm_lighthouse_assignment.go:
```diff
@@ -0,0 +1,193 @@
+	resp, err := client.Get(ctx, id.Scope, id.Name, utils.Bool(false))
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			log.Printf("[WARN] Lighthouse Assignment '%s' was not found (Scope '%s')", id.Name, id.Scope)
```
```suggestion
			log.Printf("[WARN] Lighthouse Assignment %q was not found (Scope %q)", id.Name, id.Scope)
```

Humoiz commented:
> Hi @Humoiz
> I hope you don't mind, but I've resolved a merge conflict and rebased to bring the branch up to date to help us assess if this can go in manually tested ahead of the changes we're working on for the binary testing capability that's imminently due. I'll be focused on this work today and I'm hoping we'll be able to get it in for the next release.

Thank you. I`m glad that this work is getting the traction again.

tombuildsstuff commented on azurerm/internal/services/lighthouse/resource_arm_lighthouse_definition.go:
```diff
@@ -0,0 +1,249 @@
+	resp, err := client.Get(ctx, id.Scope, id.LighthouseDefinitionID)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			log.Printf("[WARN] Lighthouse Definition '%s' was not found (Scope '%s')", id.LighthouseDefinitionID, id.Scope)
```
```suggestion
			log.Printf("[WARN] Lighthouse Definition %q was not found (Scope %q)", id.LighthouseDefinitionID, id.Scope)
```

tombuildsstuff commented on azurerm/internal/services/lighthouse/tests/resource_arm_lighthouse_assignment_test.go:
```diff
@@ -0,0 +1,216 @@
+
+		scope := rs.Primary.Attributes["scope"]
+		lighthouseAssignmentName := rs.Primary.Attributes["name"]
+		expandLighthouseDefinition := true
```
this can be `false` here to save retrieving all of the other information we don't need:

```suggestion
		expandLighthouseDefinition := false
```

jackofallops replied:
I removed that schema attribute already, and it's set to false in the API calls.  Since the tests won't work in our acceptance environment anyway I've not address the test code in the changes. We'll need to follow up when we're able to use binary testing.

jackofallops commented:
Thanks @Humoiz - I think I'm done with my changes, and I've tested all the scenarios I can think of manually, and we'll turn that into acceptance tests when we complete the binary testing work.  I'm going to get someone else from the team to double check my work, and if that gets the all clear we'll get this merged asap.

tombuildsstuff reviewed (approved):
Some minor comments whilst passing through but this otherwise LGTM 👍🏻

mbfrahry reviewed (approved):
LGTM!
